src/ir/utils.rs

```diff
@@ -225,38 +249,156 @@ impl<T, V> std::ops::Index<Idx<T>> for DenseIndexInfo<T, V> {
+        }
+    }
+
+    /// Apply a function to each component in a pre-order traversal.
+    pub fn apply_pre_order<F>(self, mut f: F)
+    where
+        F: FnMut(&Context, CompIdx),
+    {
+        for &idx in self.order.iter().rev() {
+            log::trace!("Pre-order: {}", idx);
+            f(&self.ctx, idx)
+        }
+    }
+
+    /// Take the [Context] from the post order structure.
+    pub fn take(self) -> Context {
+        self.ctx
+    }
+
+    fn process_cmd(
+        ctx: &Context,
+        comp: CompIdx,
+        cmd: &Command,
+        ts: &mut TopologicalSort<CompIdx>,
+    ) {
+        match cmd {
+            Command::Instance(inst) => {
+                let inst = ctx.comps.get(comp).instances().get(*inst);
+                // If the instance is not an external, add a dependency edge
+                if !ctx.comps.get(inst.comp).is_ext {
+                    ts.add_dependency(comp, inst.comp);
+                }
+            }
+            Command::ForLoop(fl) => {
+                for cmd in &fl.body {
+                    Traversal::process_cmd(ctx, comp, cmd, ts);
+                }
+            }
+            Command::If(i) => {
+                for cmd in &i.then {
+                    Traversal::process_cmd(ctx, comp, cmd, ts);
+                }
+                for cmd in &i.alt {
+                    Traversal::process_cmd(ctx, comp, cmd, ts);
+                }
+            }
+            Command::Connect(_) | Command::Invoke(_) | Command::Fact(_) => (),
+        }
+    }
+}
+
 #[derive(Copy)]
 /// A reference to a foreign key and its owner.
 /// On its own, a foreign key is not very useful. We need provide it with a context
 /// that can resolve the owner which can then resolve the underlying type.
 /// However, we do not provide a way to extract the underyling `T`.
-pub struct Foreign<T, C> {
+pub struct Foreign<T, C>
+where
+    C: Ctx<T>,
+{
     /// A reference to the underlying value.
     key: Idx<T>,
     /// A reference to the owner of the foreign key.
     owner: Idx<C>,
 }
 
-impl<T, C> Foreign<T, C> {
+impl<T, C> Foreign<T, C>
+where
+    C: Ctx<T>,
+{
     pub fn new(key: Idx<T>, owner: Idx<C>) -> Self {
         Self { key, owner }
     }
 
     /// Map over the foreign key using the given context.
     /// We require a context to resolve the owner of the foreign key.
-    pub fn map<X, F>(&self, mut f: F, ctx: impl Ctx<C>) -> Foreign<X, C>
+    pub fn map<X, F>(&self, ctx: &impl Ctx<C>, mut f: F) -> Foreign<X, C>
     where
-        F: FnMut(Idx<T>, &C) -> Idx<X>,
+        F: FnMut(&C, Idx<T>) -> Idx<X>,
+        C: Ctx<X>,
     {
         let c_resolved = ctx.get(self.owner);
         Foreign {
-            key: f(self.key, c_resolved),
+            key: f(c_resolved, self.key),
             owner: self.owner,
         }
     }
+
+    /// Runs a function used to unwrap the foreign type into a different type.
+    pub fn apply<F, X>(&self, ctx: &impl Ctx<C>, mut f: F) -> X
```

Comment:
Add a comment saying that you probably shouldn't use this if `X` is an `Idx<T>` type since using that index is going to be unsafe.